Web locks is not released after page navigated / refreshed

basuke opened the pull request:
#### 278f1e3a76d982a661d3cc6ea4443c1279203d9c
<pre>
Web locks is not released after page navigated / refreshed
<a href="https://bugs.webkit.org/show_bug.cgi?id=283528">https://bugs.webkit.org/show_bug.cgi?id=283528</a>
<a href="https://rdar.apple.com/140399284">rdar://140399284</a>

Reviewed by Chris Dumez.

When a page holding WebLocks enters the Back/Forward Cache (BFCache), the locks were not being released properly.
This caused locks remain held while page is in BFCache. The WebLockManager and its locks stay alive in the
suspended page

The core issue is that pages in BFCache maintain their JavaScript context and objects (including WebLockManager),
but should release shared resources like locks to avoid blocking other pages.

To solve this, override `WebLockManager::suspend()` to release locks when pages enter BFCache. Also reject all
promise before clearing from the lock manager.

Added LayoutTest to verify BFCache-specific behavior.

* LayoutTests/http/wpt/web-locks/back-forward-cache-suspend-test.https-expected.txt: Added.
* LayoutTests/http/wpt/web-locks/back-forward-cache-suspend-test.https.html: Added.
* LayoutTests/http/wpt/web-locks/resources/bfcache-helper.html: Added.
* Source/WebCore/Modules/web-locks/WebLockManager.cpp:
(WebCore::WebLockManager::clientIsGoingAway):
(WebCore::WebLockManager::suspend):
* Source/WebCore/Modules/web-locks/WebLockManager.h:

Canonical link: <a href="https://commits.webkit.org/297294@main">https://commits.webkit.org/297294@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/84084036cfcf81ffc2ba1adf5f6b7368b12ab77c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111183 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30849 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21312 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117214 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/61451 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113145 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31530 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39431 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84514 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114130 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25187 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100117 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64960 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24529 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/18258 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61034 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94565 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18325 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/120270 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38232 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28409 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93453 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38608 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96392 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93278 "Found 2 new API test failures: /WebKitGTK/TestWebKitPolicyClient:/webkit/WebKitPolicyClient/new-window-policy, /WebKitGTK/TestWebProcessExtensions:/webkit/WebKitWebView/web-process-crashed (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23773 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38364 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16135 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34217 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38121 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43598 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37786 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41119 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39488 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/web-locks/WebLockManager.cpp:
```diff
@@ -370,4 +370,20 @@ bool WebLockManager::virtualHasPendingActivity() const
     return !m_pendingRequests.isEmpty() || !m_releasePromises.isEmpty();
 }
 
+void WebLockManager::suspend(ReasonForSuspension reason)
```
Is there a way to write a test for this?

basuke replied:
I've wrote API tests. It failed without fix and pass with it.

cdumez commented on Source/WebCore/Modules/web-locks/WebLockManager.cpp:
```diff
@@ -370,4 +370,20 @@ bool WebLockManager::virtualHasPendingActivity() const
 
+void WebLockManager::suspend(ReasonForSuspension reason)
+{
+    // If we're being suspended due to page navigation or unload, force cleanup
```
Comments in WebKit end with a period.

cdumez replied:
Also the comment doesn't make a lot of sense, I wouldn't expect `suspend()` to get called on unload. `suspend()` gets called when you navigate away from the page and the page gets suspended in the back/forward cache (or when pausing in WebInspector debugging I think).

basuke replied:
Removed.

cdumez commented on Source/WebCore/Modules/web-locks/WebLockManager.cpp:
```diff
@@ -370,4 +370,20 @@ bool WebLockManager::virtualHasPendingActivity() const
+{
+    // If we're being suspended due to page navigation or unload, force cleanup
+    if (reason == ReasonForSuspension::PageWillBeSuspended || reason == ReasonForSuspension::BackForwardCache) {
+        // Force release all held locks
```
Comments in WebKit end with a period.

cdumez commented on Source/WebCore/Modules/web-locks/WebLockManager.cpp:
```diff
@@ -358,6 +358,10 @@ void WebLockManager::clientIsGoingAway()
 
+    // Reject all pending promises before clearing
+    for (auto& pair : m_releasePromises)
+        pair.value->reject(ExceptionCode::AbortError, "Lock was released due to page suspension"_s);
```
This is not only called on page suspension so this error message is not accurate I think.

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WebLocks.mm:
```diff
@@ -244,4 +244,323 @@ static void runLockRequestWaitingOnAnotherPage(ShouldUseSameProcess shouldUseSam
     runLockRequestWaitingOnAnotherPage(ShouldUseSameProcess::Yes);
 }
 
+static void runLocksReleasedOnNavigationTest(ShouldUseSameProcess shouldUseSameProcess)
```
Are the API tests really still helpful now that we have a layout test? What extra coverage do you get from API tests that is not doable with layout tests?

basuke replied:
I don't think it doesn't add extra coverage. Let's remove them.

cdumez commented on Source/WebCore/Modules/web-locks/WebLockManager.cpp:
```diff
@@ -353,11 +353,28 @@ void WebLockManager::stop()
+{
+    switch (reason) {
+    case ReasonForStop::StopIsCalled:
+        return "Promise was rejected because stop() is called"_s;
```
Those messages are way too level for Web developers. How would they know what `stop()` is? This is an internal WebKit concept. I think we should use a generic message like "Promise was rejected because the browsing context is going sway".

cdumez commented on Source/WebCore/Modules/web-locks/WebLockManager.cpp:
```diff
@@ -169,7 +169,7 @@ WebLockManager::WebLockManager(NavigatorBase& navigator)
 WebLockManager::~WebLockManager()
 {
-    clientIsGoingAway();
+    clientIsGoingAway(ReasonForStop::ContextIsGoingAway);
```
?

basuke committed:
Web locks is not released after page navigated / refreshed
https://bugs.webkit.org/show_bug.cgi?id=283528
rdar://140399284

Reviewed by Chris Dumez.

When a page holding WebLocks enters the Back/Forward Cache (BFCache), the locks were not being released properly.
This caused locks remain held while page is in BFCache. The WebLockManager and its locks stay alive in the
suspended page

The core issue is that pages in BFCache maintain their JavaScript context and objects (including WebLockManager),
but should release shared resources like locks to avoid blocking other pages.

To solve this, override `WebLockManager::suspend()` to release locks when pages enter BFCache. Also reject all
promise before clearing from the lock manager.

Added LayoutTest to verify BFCache-specific behavior.

* LayoutTests/http/wpt/web-locks/back-forward-cache-suspend-test.https-expected.txt: Added.
* LayoutTests/http/wpt/web-locks/back-forward-cache-suspend-test.https.html: Added.
* LayoutTests/http/wpt/web-locks/resources/bfcache-helper.html: Added.
* Source/WebCore/Modules/web-locks/WebLockManager.cpp:
(WebCore::WebLockManager::clientIsGoingAway):
(WebCore::WebLockManager::suspend):
* Source/WebCore/Modules/web-locks/WebLockManager.h:

Canonical link: https://commits.webkit.org/297294@main
